internal/subscription/delay.go

```diff
@@ -0,0 +1,199 @@
+
+// DelayConsumerConfig ...
+type DelayConsumerConfig struct {
+	Topic           string `json:"topic"`
+	GroupID         string `json:"group_id"`
+	GroupInstanceID string `json:"group_instance_id"`
+}
+
+// LogFields ...
+func (dc DelayConsumerConfig) LogFields(kv ...interface{}) []interface{} {
+	fields := []interface{}{
+		"delayConsumerConfig", map[string]interface{}{
+			"topic":           dc.Topic,
+			"groupID":         dc.GroupID,
+			"groupInstanceID": dc.GroupInstanceID,
+		},
+	}
+
+	fields = append(fields, kv...)
+	return fields
+}
+
+// DelayConfig ...
+type DelayConfig struct {
+	MinimumBackoffInSeconds uint     `json:"minimum_backoff_in_seconds"`
+	MaximumBackoffInSeconds uint     `json:"maximum_backoff_in_seconds"`
+	MaxDeliveryAttempts     int32    `json:"max_delivery_attempts"`
+	Subscription            string   `json:"subscription"`
+	DeadLetterTopic         string   `json:"dead_letter_topic"`
+	DelayTopics             []string `json:"delay_topics"`
+	// maintains a mapping for all the delay intervals to its delay topic name. This will be used to lookup the next topic name.
+	DelayIntervalToTopicMap map[Interval]DelayConsumerConfig `json:"delay_interval_to_topic_map"`
+}
+
+// NewDelayConfig validates a subscription model and initializes the needed config values to be used for delay consumers
+func NewDelayConfig(model *Model) (*DelayConfig, error) {
+
+	minB := defaultMinimumBackoffInSeconds
+	maxB := defaultMaximumBackoffInSeconds
+	maxAtt := defaultMaxDeliveryAttempts
+
+	if model.RetryPolicy != nil {
+		minB = model.RetryPolicy.MinimumBackoff
+		if minB < lowerBoundMinimumBackoffInSeconds || minB > upperBoundMinimumBackoffInSeconds {
+			return nil, ErrInvalidMinBackoff
+		}
+	}
+
+	if model.RetryPolicy != nil {
+		maxB = model.RetryPolicy.MaximumBackoff
+		if maxB < lowerBoundMaximumBackoffInSeconds || maxB > upperBoundMaximumBackoffInSeconds {
+			return nil, ErrInvalidMaxBackoff
+		}
+	}
+
+	// currently dl-topics are auto-created for subscriptions, refer GetValidatedModelForCreate()
+	if model.DeadLetterPolicy != nil {
+		maxAtt = model.DeadLetterPolicy.MaxDeliveryAttempts
+		if maxAtt == 0 {
+			// since dl-topics are auto-created, a user may never provide a dead-letter policy during subscription creation.
+			// in such cases we set MaxDeliveryAttempts to a default value
+			maxAtt = defaultMaxDeliveryAttempts
+		} else if maxAtt < lowerBoundMaxDeliveryAttempts || maxAtt > upperBoundMaxDeliveryAttempts {
+			return nil, ErrInvalidMaxDeliveryAttempt
+		}
+	}
+
+	delayTopics := make([]string, 0)
+	delayIntervalToTopicNameMap := make(map[Interval]DelayConsumerConfig)
+	groupID := uuid.New().String()
+	for _, interval := range Intervals {
+		delayTopic := topic.GetTopicName(model.ExtractedSubscriptionProjectID, fmt.Sprintf(delayTopicNameFormat, model.ExtractedSubscriptionName, interval))
+		delayTopics = append(delayTopics, delayTopic)
+		delayIntervalToTopicNameMap[interval] = DelayConsumerConfig{
```

Comment:
DelayConsumer logic should be handled by subscriber

Reply:
Agreed. There is no logic here. Its just converting a slice of topic names to map for quick lookups.